### PR TITLE
[wip] magit-explain-section

### PR DESCRIPTION
**Pull request by npostavs**
This is my initial attempt at #3308. It seems to work okay, though the insertion function docstrings are not written with this exact use case in mind, so the `*Help*` screen reads a bit awkward.

I wasn't sure whether we should take over the `magit-describe-section` name from the current command, so I've called the new command `magit-explain-section` for now. I haven't added a keybinding yet, though I was thinking perhaps <kbd>? ?</kbd>.

I have a couple question in the code, which I'll call out through inline comments on the PR.

Any comments in general are welcome of course.

### CODE REVIEW

**Comment by tarsius** [comment]
> I wasn't sure whether we should take over the `magit-describe-section` name from the current command,

We could rename the existing command to `magit-describe-section-briefly` and then use the old name for the new command, analogous to `describe-key{,-briefly}`.

File: lisp/magit-mode.el [hunk 1]
```diff
@@ -599,7 +599,7 @@ locked to its value, which is derived from MODE and ARGS."
       (funcall mode))
     (magit-display-buffer buffer)
     (with-current-buffer buffer
-      (run-hooks 'magit-mode-setup-hook)
+      (run-hooks 'magit-mode-setup-hook) ; run-mode-hooks?
```
[the remark this answers is not in the document]
Comment (tarsius) [reply lisp/magit-mode.el]: What makes you think so?

The mode hooks were already run three lines above with `(funcall mode)`.

[the remark this answers is not in the document]
Comment (npostavs) [reply lisp/magit-mode.el]: Oh, you're right, I was just confused.

**Comment by npostavs** [comment]
> We could rename the existing command to magit-describe-section-briefly and then use the old name for the new command, analogous to describe-key{,-briefly}.

I like this idea.

File: lisp/magit-section.el [hunk 2]
```diff
@@ -881,29 +927,32 @@ insert a newline character if necessary."
...
+    (magit-run-section-hook hook)
+    (when header-sections
+      ;; Make the first header into the parent of the rest.
+      (cl-callf nreverse header-sections)
+      (let* ((1st-header (pop header-sections))
+             (header-parent (oref 1st-header parent)))
+        ;; Can we just assume the header sections are the only
+        ;; sections, and then we don't need the collector function in
+        ;; `magit-insert-section-hook' at all?
+        (cl-assert (equal (oref header-parent children) (cons 1st-header header-sections)))
```
[the remark this answers is not in the document]
Comment (tarsius) [reply lisp/magit-section.el]: > Can we just assume the header sections are the only sections,

As in; can we assume they are the first sections? I guess currently we can, but why impose such a restriction?

File: lisp/magit-section.el [hunk 3]
```diff
@@ -175,8 +175,12 @@ but in the future it will also be used set the defaults."
 
 ;;; Core
 
+(defvar magit--current-section-hook nil
+  "Internal variable used for `magit-explain-section'.")
+
 (defclass magit-section ()
-  ((type     :initform nil :initarg :type)
+  ((source   :initform (symbol-value 'magit--current-section-hook))
+   (type     :initform nil :initarg :type)
```
Comment (tarsius) [inline lisp/magit-section.el]: `type` should be first because... it sets the tone. Actually it shouldn't exist at all, we should use more classes instead.

That's just a preference, but `source` should appear further down; I would place it before `parent`. It also shouldn't be named `source` - `magit-file-section` already adds a slot by that name for a different purpose. Maybe use a name containing "hook".

**Comment by npostavs** [comment]
> The current implementation tries to insert one header until it succeeds in doing so, i.e. inserting the section is not canceled. Once that has happened it inserts the remaining headers as children of that section. Does your impementation do that too?

My code inserts all the sections, and then makes the first section into the parent of the remaining ones. I believe this is equivalent.

> Also do we actually need to do this to make your other changes work for these sections too?

The previous implementation is spread across a couple of hooks, which I found really hard to follow. I don't think I could keep it in my head at the same time as writing my other changes (so I'm unsure whether my other changes are compatible with the original code or not).

**Comment by tarsius** [comment]
Alright, lets go with your approach.

File: lisp/magit-section.el (hunk 2 above)
[the remark this answers is not in the document]
Comment (npostavs) [reply lisp/magit-section.el]: > why impose such a restriction?

It would this code slightly simpler; but since you're asking the question, that means this means assumption is only satisfied by accident, and we shouldn't rely on it. I'll just remove the assert.

File: lisp/magit-section.el (hunk 3 above)
Comment (npostavs) [reply lisp/magit-section.el]: Sure, I just put it at the top because I didn't know where to put it, and it's conveniently closer to the `defvar`.

> It also shouldn't be named `source` - `magit-file-section` already adds a slot by that name for a different purpose.

Oh, I missed that.
Maybe `origin-hook`, `creating-hook`, `inserter`, `made-by`?  Ugh, naming is hard.

**Comment by tarsius** [comment]
Do you need anything else from me? You've got my okay to merge this.

**Comment by npostavs** [comment]
> Do you need anything else from me?

No, it's just a matter of finding free time. Rebased in #3539.

----

By the way, about this:

> I haven't added a keybinding yet, though I was thinking perhaps <kbd>? ?</kbd>.

Turns out this conflicts with the existing popup help chars, so I'm going to just leave it unbound I think.
